File: python/sort/selection_sort.py
'''Selection Sort

1. Divide input list into two sublists: sorted list and unsorted list
2. From unsorted list, find the minimum and exchange minimum with leftmost number
3. Move the sublist boundaries one element to the right'''


def select_sort(arr, reverse = False):
	'''Sort given list with selection sort algorithm'''

	arr = [[], arr]

	# Sort Descending
	if reverse == True:
		while arr[1]:
			for i in range(len(arr[1])):
				if arr[1][i] == max(arr[1]):
					arr[1][0], arr[1][i] = arr[1][i], arr[1][0]
			arr[0].append(arr[1].pop((0)))

	# Sort Ascending
	else:
		while arr[1]:
			for i in range(len(arr[1])):
				if arr[1][i] == min(arr[1]):
					arr[1][0], arr[1][i] = arr[1][i], arr[1][0]
			arr[0].append(arr[1].pop(0))

	return arr[0]


def select_hard(arr):
	"""Selection sort with in given list arr"""

	length = len(arr)

	for i in range(length):
		min_index = i
		for j in range(i, length):
			if arr[j] < arr[min_index]:
				min_index = j
		arr[i], arr[min_index] = arr[min_index], arr[i]

	return arr

if __name__ == "__main__":
	test = [int(i) for i in input().split(',')]
	
	print(select_hard(test))
	print(test)
	print(select_sort(test))
	print(test)
	

File: javascript/sort/selectionSort.js
/* Selection Sort 

Sort given array by repeatedly finding the minimum element 
(considering ascending order) from unsorted part 
and putting it at the beginning. 

selectionSort(array[, reverse])
    * array: target array to sort
    * reverse: take boolean value. optional param. 
            true for descending, false for ascending.
*/

function selectionSort(arr, reverse = false) {
  function swap(i, j) {
    // Swap arr[i] with arr[j]
    const tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
  }

  const n = arr.length;

  // Sort
  for (let i = 0; i < n - 1; i++) {
    // if reverse, idx represents index of max in arr.
    // if not reverse, idx represents index of min in arr.
    let idx = i;

    for (let j = i + 1; j < n; j++) {
      // if reverse, find max from remaining elements and assign index to idx
      if (reverse === false && arr[idx] > arr[j]) idx = j;
      // if not reverse, assign index of min to idx
      else if (reverse && arr[idx] < arr[j]) idx = j;
    }

    // if idx !== i, it means arr[i] need to be taken over by arr[idx]
    if (idx !== i) swap(idx, i);
  }

  return arr;
}
